src/ctapipe/image/modifications.py

```diff
@@ -100,6 +112,256 @@ def _smear_psf_randomly(
+                    nsb_database[tel_id].append(r1.waveform)
+
+        nsb_database = {
+            tel_id: np.stack(waveforms) for tel_id, waveforms in nsb_database.items()
+        }
+
+        # Check that we have enough NSB-only events for all telescopes. We
+        # require that the number of NSB events for any telescope is at
+        # least two times the number of waveforms (=nsb_level) that we will add
+        # up. This is to avoid excessive correlation among the waveforms.
+        stats_ok = True
+        for tel_id in nsb_database:
+            nevents = nsb_database[tel_id].shape[0]
+            if nevents >= 2 * self.nsb_level:
+                continue
+            self.log.error(
+                f"Not enough NSB events available for tel_"
+                f"id {tel_id}. "
+                f"For nsb_level = {self.nsb_level}, at least "
+                f"{2*self.nsb_level} events are needed ({nevents} "
+                f"were found)."
+            )
+            stats_ok = False
+
+        if not stats_ok:
+            raise ValueError("Please use an input nsb_file with more events!")
+
+        # Now shift the waveforms so that they have mean=0 and do not introduce
+        # any bias (just fluctuations). For each telescope and gain we average
+        # all pixels
+        for tel_id in nsb_database:
+            for channel in range(nsb_database[tel_id].shape[1]):
+                mean = np.mean(nsb_database[tel_id][:, channel, :, :])
+                nsb_database[tel_id][:, channel, :, :] -= mean
+
+        # Now add up waveforms selected at random to obtain different
+        # realizations of the total noise that will be added
+        for tel_id in nsb_database:
+            self.total_noise[tel_id] = build_wf_noise_pixelwise(
+                nsb_database[tel_id],
+                self.n_noise_realizations,
+                self.nsb_level,
+                self.rng,
+                self.shuffle_full_cameras,
+            )
+
+    def __call__(self, tel_id, waveforms, selected_gain_channel=None):
+        """
+
+        Parameters
+        ----------
+        tel_id
+        waveforms: ndarray [ngains, npixels, nsamples] (ngains=1 if
+        gain-selected)
+        selected_gain_channel: ndarray[npixels] or None if no gain selection
+
+        Returns
+        -------
+        ndarray, same shape as waveforms: original waveforms plus added NSB
+
+        """
+
+        # Note: MC waveforms passed to this function should contain data in all
+        # pixels (not DVR'ed - obviously DVR depends on noise level, it does not
+        # make sense to add the noise after DVR was applied)
+        if selected_gain_channel is None:
+            return (
+                waveforms
+                + self.total_noise[tel_id][self.rng.integers(self.n_noise_realizations)]
+            )
+        else:
+            return (
+                waveforms
+                + self.total_noise[tel_id][
```

Comment:
This if else is a bit hard to read and see the differences.

Maybe it would be simpler like this:

```python
if selected_gain_channel is None:
    noise = ...
else:
   noise = ...

return waveforms + noise
```